src/hist/basehist.py

```diff
@@ -150,20 +161,140 @@ def project(self, *args: Union[int, str]):
+            a Hist object.
+        """
+
+        if len(args) and not len(kwargs):
+            return super().fill(*args)
+
+        elif (
+            isinstance(kwargs, dict)
+            and all(isinstance(k, str) for k in kwargs.keys())
+            and not len(args)
+        ):
+            indices: dict = {}
+            for n, v in kwargs.items():
+                indices[self._name_to_index(n)] = v
+
+            lst = sorted(indices.items(), key=lambda item: item[0])
+            nd = np.asarray(lst, dtype=object)
+            data = nd.ravel()[1::2]
+            return super().fill(*data)
+
+        else:
+            raise TypeError(
+                f"Only dict with keys of int or str is supported for {self.__class__.__name__}"
             )
 
+    def _loc_shortcut(self, x):
+        """
+            Convert some specific indices to location.
+        """
+
+        if isinstance(x, slice):
+            return slice(
+                self._loc_shortcut(x.start),
+                self._loc_shortcut(x.stop),
+                self._step_shortcut(x.step),
+            )
+        elif isinstance(x, complex):
+            if x.real % 1 != 0:
+                raise ValueError("The real part should be an integer")
+            else:
+                return bh.loc(x.imag, int(x.real))
+        elif isinstance(x, str):
+            return bh.loc(x)
+        else:
+            return x
+
+    def _step_shortcut(self, x):
+        """
+            Convert some specific indices to step.
+        """
+
+        if isinstance(x, complex):
+            if x.real != 0:
+                raise ValueError("The step should not have real part")
+            elif x.imag % 1 != 0:
+                raise ValueError("The imaginary part should be an integer")
+            else:
+                return bh.rebin(int(x.imag))
+        else:
+            return x
+
+    def __getitem__(self, index):
+        """
+            Get histogram item.
+        """
+
+        if isinstance(index, dict):
+
+            if all(isinstance(k, int) for k in index.keys()):
+                return super().__getitem__(
+                    {k: self._loc_shortcut(v) for k, v in index.items()}
+                )
+
+            elif all(isinstance(k, str) for k in index.keys()):
+                indices: dict = {}
+                for n, v in list(index.items()):
+                    indices[self._name_to_index(n)] = self._loc_shortcut(v)
+
+                return super().__getitem__(indices)
+
+            else:
+                raise TypeError(
+                    f"Only dict with keys of int or str is supported for {self.__class__.__name__}"
+                )
+
+        if not hasattr(index, "__iter__"):
+            index = (index,)
+
+        return super().__getitem__(tuple(self._loc_shortcut(v) for v in index))
+
+    def __setitem__(self, index, value):
+        """
+            Set histogram item.
+        """
+
+        if isinstance(index, dict):
```

Comment:
You can factor out the index transform logic into a hidden function if you want, it should be the same for these two functions (set/get).

Reply:
Yep, I have thought about this previously. But it's a little bit tricky, as `return super().__set/getittem__` are nested in the logic.

Reply:
Once you rewrite this to simply transform "index" with 0 or more string into an "index" with only ints, then it will be be disconnected from the logic and will be easy to factor out.